src/next/middleware.ts

```diff
@@ -0,0 +1,162 @@
+   * @default process.env.NEXTAUTH_SECRET
+   *
+   * ---
+   * [Documentation](https://next-auth.js.org/configuration/options#secret)
+   */
+  secret?: string
+  /**
+   * Where to redirect the user in case of an error if they weren't logged in.
+   * Similar to `pages` in `NextAuth`.
+   *
+   * ---
+   * [Documentation](https://next-auth.js.org/configuration/pages)
+   */
+  pages?: NextAuthOptions["pages"]
+  /**
+   * Callback that receives the user's JWT payload
+   * and returns `true` to allow the user to continue.
+   *
+   * If it returns `false`, the user is redirected to the sign-in page instead
+   *
+   * The default is to let the user continue if they have a valid JWT (basic authentication).
+   *
+   * How to restrict a page and all of it's subpages for admins-only:
+   * @example
+   *
+   * ```js
+   * // `pages/admin/_middleware.js`
+   * import { withAuth } from "next-auth/next/middleware"
+   *
+   * export default withAuth({
+   *   authorized: ({ token }) => token?.user.isAdmin
+   * })
+   * ```
+   */
+  authorized?: (options: {
+    token: JWT | null
+    req: NextRequest
+  }) => Awaitable<boolean>
+}
+
+export type WithAuthArgs =
+  | [NextRequest]
+  | [NextRequest, NextFetchEvent]
+  | [NextRequest, NextAuthMiddlewareOptions]
+  | [NextMiddleware]
+  | [NextMiddleware, NextAuthMiddlewareOptions]
+  | [NextAuthMiddlewareOptions]
+
+/** Check if `secret` has been declared  */
+function initConfig(
+  req: NextRequest,
+  optionsOrEv?: NextAuthMiddlewareOptions | NextFetchEvent
+) {
+  const options = optionsOrEv as NextAuthMiddlewareOptions // This is relatively safe as we don't share properties with NextFetchEvent
+  const signInPage = options?.pages?.signIn ?? "/api/auth/signin"
+  const errorPage = options?.pages?.error ?? "/api/auth/error"
+
+  // Avoid infinite redirect loop
+  if ([signInPage, errorPage].includes(req.nextUrl.pathname)) return
+
+  const secret = options?.secret ?? process.env.NEXTAUTH_SECRET
+  // Continue only if the secret is specified
+  if (secret)
+    return {
+      req,
+      options,
+      secret,
+      signInPage,
+      authorized: options?.authorized ?? (({ token }) => token),
+    }
+
+  console.error(
+    `[next-auth][error][NO_SECRET]`,
+    `\nhttps://next-auth.js.org/errors#no_secret`
+  )
+
+  return {
+    redirect: NextResponse.redirect(`${errorPage}?error=Configuration`),
+  }
+}
+
+/**
+ * Middleware that checks if the user is authenticated/authorized.
+ * If if they aren't, they will be redirected to the login page.
+ * Otherwise, continue.
+ *
+ * @example
+ *
+ * ```js
+ * // `pages/_middleware.js`
+ * export { default } from "next-auth/middleware"
+ * ```
+ *
+ * ---
+ * [Documentation](https://next-auth.js.org/getting-started/middleware)
+ */
+export function withAuth(...args: WithAuthArgs) {
+  if (args[0] instanceof NextRequest) {
+    const config = initConfig(args[0], args[1])
+    if (!config || config.redirect) return config?.redirect
```

Comment:
I see two flows here, either we want to **(1)** redirect because there was an error in the configuration or **(2)** or do nothing to prevent the infinite loop.

I suggest to re-write this bit to be more explicit:
```ts
// if there was an error reading the middleware config redirect to the error page
if (config.errorRedirect)  return config.errorRedirect
// prevent an infinite loop when coming from an OAuth flow like `/api/auth/callback/{provider}`
if (!config) return
```
so it's easier to grasp at first glance what's happening here without having to look at the implementation of `initConfig()`